Source/Particles/WarpXParticleContainer.cpp

```diff
@@ -593,7 +593,246 @@ WarpXParticleContainer::DepositCharge (WarpXParIter& pti, RealVector const& wp,
                                        const long offset, const long np_to_depose,
                                        int thread_num, int lev, int depos_lev)
 {
-    if (!do_not_deposit) {
+    // If user decides not to deposit
+    if (do_not_deposit) return;
+
+    if (WarpX::do_shared_mem_charge_deposition)
```

Comment:
A number of tests and checks are done here within `if (WarpX::do_shared_mem_charge_deposition)`. However, I think that they are not specific to the shared memory implementation. Intuitively, the `if (WarpX::do_shared_mem_charge_deposition)` should only contain: 
- The creation of the `bins` object
- The call to `doChargeDepositionSharedShapeN` 
and the rest should be common to both deposition.
Would that make sense?

Reply:
A lot of the code that would be factored out as common got moved to the ABLASTR implementation. I'll try pulling up what I can of what's left, though.